keep-runtime/src/workload.rs

```diff
@@ -0,0 +1,150 @@
+    /// import field not found
+    ImportFieldNotFound(String, String),
+    /// export not found
+    ExportNotFound,
+    /// call failed
+    CallFailed,
+    /// runtime error
+    RuntimeError,
+    /// I/O error
+    IoError(std::io::Error),
+}
+
+impl From<std::io::Error> for Error {
+    fn from(err: std::io::Error) -> Self {
+        Self::IoError(err)
+    }
+}
+
+/// Result type used throughout the library.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Runs a WebAssembly workload.
+pub fn run(
+    reader: &mut impl std::io::Read,
+    args: Box<[String]>,
+    envs: Box<[(String, String)]>,
+) -> Result<Box<[wasmtime::Val]>> {
+    let mut buf = Vec::new();
+    reader.read_to_end(&mut buf)?;
+
+    let engine = wasmtime::Engine::default();
+    let store = wasmtime::Store::new(&engine);
+
+    // Instantiate WASI.  We enable `wasi_snapshot_preview1` and
+    // `wasi_unstable` for compatibility reasons.
+    let mut builder = wasi_common::WasiCtxBuilder::new();
+    builder.args(&args.to_vec()).envs(&envs.to_vec());
+    let ctx = builder.build().or(Err(Error::RuntimeError))?;
+    let wasi_snapshot_preview1 = wasmtime_wasi::Wasi::new(&store, ctx);
+
+    let mut builder = wasi_common::old::snapshot_0::WasiCtxBuilder::new();
+    builder.args(&args.to_vec()).envs(&envs.to_vec());
+    let ctx = builder.build().or(Err(Error::RuntimeError))?;
+    let wasi_unstable = wasmtime_wasi::old::snapshot_0::Wasi::new(&store, ctx);
+
+    let instance = {
+        let module =
+            wasmtime::Module::new(&store, buf.into_boxed_slice()).or(Err(Error::RuntimeError))?;
+        let imports = module
+            .imports()
+            .map(|import| {
+                let module_name = import.module();
+                let field_name = import.name();
+                let export = match module_name {
+                    "wasi_snapshot_preview1" => Ok(wasi_snapshot_preview1.get_export(field_name)),
+                    "wasi_unstable" => Ok(wasi_unstable.get_export(field_name)),
+                    _ => Err(Error::ImportModuleNotFound(module_name.to_string())),
+                }?;
+
+                if let Some(export) = export {
+                    Ok(export.clone().into())
+                } else {
+                    Err(Error::ImportFieldNotFound(
+                        module_name.to_string(),
+                        field_name.to_string(),
+                    ))
+                }
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        wasmtime::Instance::new(&module, &imports).or(Err(Error::RuntimeError))?
+    };
+
+    let function = if instance.exports().any(|export| export.name().is_empty()) {
+        // Launch the default command export.
+        instance.get_func("")
+    } else {
+        // If the module doesn't have a default command
+        // export, launch the _start function if one is
+        // present, as a compatibility measure.
+        instance.get_func("_start")
```

Comment:
This claims to maintain compatibility with something. What? Unless I'm missing something, we're writing new code and don't have to maintain compatibility with anything.

Reply:
The logic was copied from wasmtime, and even the latest toolchain seems to emit `_start` instead of the default export. I guess we need to keep it for some time.